File: oxidize-golang/core/video/frame_sampler.go
package video

import "sort"

// SampleIndices picks frame indices from [0, totalFrames) using strategy.
func SampleIndices(totalFrames, targetFrames int, strategy FrameSamplingStrategy) ([]int, error) {
	if totalFrames <= 0 || targetFrames <= 0 {
		return nil, ErrFrameCountOutRange
	}
	var indices []int
	switch strategy {
	case SampleDense:
		indices = dense(totalFrames, targetFrames, 1)
	default:
		indices = uniform(totalFrames, targetFrames)
	}
	if len(indices) == 0 {
		return nil, ErrEmptySample
	}
	return indices, nil
}

// LumaHistogramRGB builds a 16-bin normalized luma histogram for an RGB frame.
func LumaHistogramRGB(data []byte) []float32 {
	hist := make([]float32, 16)
	if len(data) == 0 {
		return hist
	}
	var total float32
	for i := 0; i+2 < len(data); i += 3 {
		luma := 0.299*float32(data[i]) + 0.587*float32(data[i+1]) + 0.114*float32(data[i+2])
		bin := int(luma / 16)
		if bin > 15 {
			bin = 15
		}
		hist[bin]++
		total++
	}
	if total > 0 {
		for i := range hist {
			hist[i] /= total
		}
	}
	return hist
}

// SampleIndicesAdaptive keeps first/last frames and fills remaining slots by
// histogram distance. Falls back to uniform when lumaHists is too short.
func SampleIndicesAdaptive(totalFrames, targetFrames int, lumaHists []float32) ([]int, error) {
	if totalFrames <= 0 || targetFrames <= 0 {
		return nil, ErrFrameCountOutRange
	}
	if len(lumaHists) < totalFrames*16 {
		return SampleIndices(totalFrames, targetFrames, SampleAdaptive)
	}
	if totalFrames <= targetFrames {
		out := make([]int, totalFrames)
		for i := range out {
			out[i] = i
		}
		return out, nil
	}
	chosen := map[int]struct{}{0: {}, totalFrames - 1: {}}
	out := []int{0, totalFrames - 1}
	for len(out) < targetFrames {
		bestIdx := -1
		var bestScore float32
		for cand := 0; cand < totalFrames; cand++ {
			if _, ok := chosen[cand]; ok {
				continue
			}
			score := minHistDistance(cand, out, lumaHists)
			if bestIdx < 0 || score > bestScore {
				bestIdx = cand
				bestScore = score
			}
		}
		if bestIdx < 0 {
			break
		}
		chosen[bestIdx] = struct{}{}
		out = append(out, bestIdx)
	}
	sort.Ints(out)
	if len(out) == 0 {
		return nil, ErrEmptySample
	}
	return out, nil
}

func uniform(total, target int) []int {
	if total <= target {
		out := make([]int, total)
		for i := range out {
			out[i] = i
		}
		return out
	}
	step := float64(total-1) / float64(target-1)
	out := make([]int, 0, target)
	seen := map[int]struct{}{}
	for i := 0; i < target; i++ {
		idx := int(float64(i)*step + 0.5)
		if idx >= total {
			idx = total - 1
		}
		if _, ok := seen[idx]; !ok {
			seen[idx] = struct{}{}
			out = append(out, idx)
		}
	}
	sort.Ints(out)
	return out
}

func dense(total, target, stride int) []int {
	if stride <= 0 {
		stride = 1
	}
	out := make([]int, 0, target)
	for i := 0; i < total && len(out) < target; i += stride {
		out = append(out, i)
	}
	return out
}

func minHistDistance(cand int, chosen []int, hists []float32) float32 {
	candHist := hists[cand*16 : (cand+1)*16]
	var best float32
	for _, idx := range chosen {
		other := hists[idx*16 : (idx+1)*16]
		d := l1(candHist, other)
		if best == 0 || d < best {
			best = d
		}
	}
	return best
}

func l1(a, b []float32) float32 {
	var s float32
	for i := range a {
		d := a[i] - b[i]
		if d < 0 {
			d = -d
		}
		s += d
	}
	return s
}

File: oxidize-python/oxidize_python/core/video/frame_sampler.py
"""Frame sampling mirroring oxidize-golang/core/video/frame_sampler.go."""

from __future__ import annotations

from enum import IntEnum

from oxidize_python.core.video.errors import (
    EmptySampleError,
    FrameCountOutOfRangeError,
)


class FrameSamplingStrategy(IntEnum):
    UNIFORM = 0
    DENSE = 1
    ADAPTIVE = 2


def sample_indices(
    total_frames: int,
    target_frames: int,
    strategy: FrameSamplingStrategy = FrameSamplingStrategy.UNIFORM,
) -> list[int]:
    """Pick frame indices from [0, total_frames) using strategy."""
    if total_frames <= 0 or target_frames <= 0:
        raise FrameCountOutOfRangeError()
    if strategy == FrameSamplingStrategy.DENSE:
        indices = _dense(total_frames, target_frames, 1)
    else:
        indices = _uniform(total_frames, target_frames)
    if not indices:
        raise EmptySampleError()
    return indices


def luma_histogram_rgb(data: bytes) -> list[float]:
    """Build a 16-bin normalized luma histogram for an RGB frame."""
    hist = [0.0] * 16
    total = 0.0
    for i in range(0, len(data) - 2, 3):
        luma = 0.299 * data[i] + 0.587 * data[i + 1] + 0.114 * data[i + 2]
        bin_idx = min(15, int(luma / 16))
        hist[bin_idx] += 1.0
        total += 1.0
    if total:
        hist = [v / total for v in hist]
    return hist


def sample_indices_adaptive(
    total_frames: int,
    target_frames: int,
    luma_hists: list[float],
) -> list[int]:
    """Keep first/last frames and fill remaining slots by histogram distance.

    `luma_hists` is a flattened array of `total_frames` 16-bin histograms
    (row-major). Falls back to uniform sampling when it is too short.
    """
    if total_frames <= 0 or target_frames <= 0:
        raise FrameCountOutOfRangeError()
    if len(luma_hists) < total_frames * 16:
        return sample_indices(total_frames, target_frames, FrameSamplingStrategy.ADAPTIVE)
    if total_frames <= target_frames:
        return list(range(total_frames))

    chosen: set[int] = {0, total_frames - 1}
    out: list[int] = [0, total_frames - 1]
    while len(out) < target_frames:
        best_idx = -1
        best_score = 0.0
        for cand in range(total_frames):
            if cand in chosen:
                continue
            score = _min_hist_distance(cand, out, luma_hists)
            if best_idx < 0 or score > best_score:
                best_idx = cand
                best_score = score
        if best_idx < 0:
            break
        chosen.add(best_idx)
        out.append(best_idx)
    out.sort()
    if not out:
        raise EmptySampleError()
    return out


def _uniform(total: int, target: int) -> list[int]:
    if total <= target:
        return list(range(total))
    step = (total - 1) / max(target - 1, 1)
    out: list[int] = []
    seen: set[int] = set()
    for i in range(target):
        idx = min(total - 1, int(i * step + 0.5))
        if idx not in seen:
            seen.add(idx)
            out.append(idx)
    out.sort()
    return out


def _dense(total: int, target: int, stride: int) -> list[int]:
    if stride <= 0:
        stride = 1
    out: list[int] = []
    i = 0
    while i < total and len(out) < target:
        out.append(i)
        i += stride
    return out


def _min_hist_distance(cand: int, chosen: list[int], hists: list[float]) -> float:
    cand_hist = hists[cand * 16 : (cand + 1) * 16]
    best = 0.0
    for idx in chosen:
        other = hists[idx * 16 : (idx + 1) * 16]
        d = _l1(cand_hist, other)
        if best == 0.0 or d < best:
            best = d
    return best


def _l1(a: list[float], b: list[float]) -> float:
    return sum(abs(x - y) for x, y in zip(a, b))
